### PR TITLE
Bug tojson memleak

### LINKED ISSUE #43877: BUG: to_json memory leak (introduced in 1.1.0)
### 

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the master branch of pandas.

### Reproducible Example

```python
import pandas as pd
import numpy as np

# any loop
for _ in range(1000)
    df = pd.DataFrame({str(c): np.random.random_sample(size=100_000) for c in range(10)})
    df.to_json() # same regardless orient or using file
```

### Issue Description

It looks like a memory leak when calling `to_json` introduced in version 1.1.0. It seems it prevents the dataframe to be correctly garbage collected. Here's a memory profile Pandas 1.1.0 compared to the previous version 1.0.5:

mem_prof

I have the same trends either on Windows 10, Linux Ubuntu, python 3.7, 3.8 & 3.9.
This leak is still there on latest Pandas version 1.3.3 and is proportional to the size of the dataframe. I've tried direct calls to `del` and `gc.collect()` but it doesn't change anything.

It's specific to `to_json` method. I haven't observed leak with other formats such as CSV.

I don't know if it makes sense or help, here's an output using tracemalloc from this code:

```
def foo():
    df = pd.DataFrame({str(c): np.random.random_sample(size=100_000) for c in range(5)})
    df.to_json()


if __name__ == "__main__":
    tracemalloc.start(50)

    foo()

    snapshot = tracemalloc.take_snapshot()
    top_stats = snapshot.statistics('traceback')

    # pick the biggest memory block
    stat = top_stats[0]
    print("%s memory blocks: %.1f KiB" % (stat.count, stat.size / 1024))
    for line in stat.traceback.format():
        print(line)
```

with Pandas 1.1.0 or 1.3.3:

```
5 memory blocks: 782.5 KiB
  File "main.py"
    foo()
  File "main.py"
    df.to_json()
  File ".\ven37\lib\site-packages\pandas\core\generic.py", line 2571
    storage_options=storage_options,
  File ".\ven37\lib\site-packages\pandas\io\json\_json.py", line 122
    indent=indent,
  File ".\ven37\lib\site-packages\pandas\io\json\_json.py", line 183
    indent=self.indent,
  File ".\ven37\lib\site-packages\pandas\core\indexes\base.py", line 4367
    return self._data
  File ".\ven37\lib\site-packages\pandas\core\indexes\range.py", line 186
    return np.arange(self.start, self.stop, self.step, dtype=np.int64)
```

whereas 1.0.5 produces this:

```
9 memory blocks: 1.6 KiB
  File "main.py"
    foo()
  File "main.py"
    df.to_json()
  File ".\ven37\lib\site-packages\pandas\core\generic.py", line 2364
    indent=indent,
  File ".\ven37\lib\site-packages\pandas\io\json\_json.py", line 85
    indent=indent,
  File ".\ven37\lib\site-packages\pandas\io\json\_json.py", line 145
    self.indent,
  File ".\ven37\lib\site-packages\pandas\io\json\_json.py", line 245
    indent,
  File ".\ven37\lib\site-packages\pandas\io\json\_json.py", line 167
    indent=indent,
```

### Expected Behavior

No leak expected, similar to version 1.0.5

### Installed Versions

**Versions with leak:**
master -----------------
commit           : 65998341034861527517756e6dd134dafea2b766
python           : 3.8.5.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
Version          : 10.0.19041
machine          : AMD64
processor        : Intel64 Family 6 Model 158 Stepping 13, GenuineIntel
pandas           : 1.4.0.dev0+833.g6599834103
numpy            : 1.21.2
pytz             : 2021.1
dateutil         : 2.8.2
...

1.3.3 ------------------
commit           : 73c68257545b5f8530b7044f56647bd2db92e2ba
python           : 3.7.6.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
Version          : 10.0.19041
machine          : AMD64
processor        : Intel64 Family 6 Model 158 Stepping 13, GenuineIntel
pandas           : 1.3.3
numpy            : 1.21.2
...

1.1.0 ------------------
commit           : d9fff2792bf16178d4e450fe7384244e50635733
pandas           : 1.1.0
numpy            : 1.21.2
...

**Versions without leak:**
commit           : None
pandas           : 1.0.5
numpy            : 1.21.2
...

### PR DESCRIPTION
**Pull request by vernetya**
- [x] closes #43877
- [x] tests passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [x] whatsnew entry

Fix memory leak when calling `to_json` appeared in 1.1.0. From these [changes](https://github.com/pandas-dev/pandas/commit/402c5cdd580a25699f6b8b0bade22b9eb0282d08), in `get_values` function, it looks like `values` from [here ](https://github.com/pandas-dev/pandas/blob/main/pandas/_libs/src/ujson/python/objToJSON.c#L224) are not cleaned in this [path](https://github.com/pandas-dev/pandas/blob/main/pandas/_libs/src/ujson/python/objToJSON.c#L231).

I run the following loop and plot the memory usage using [memory_profiler](https://pypi.org/project/memory-profiler/):

``` python
def foo():
    data = {str(c): list(range(100_000)) for c in range(10)}
    for i in range(500):
        print(i)
        df = pd.DataFrame(data)
        df.to_json(orient='split', indent=0)
```

Here on linux (python 3.8.10)
pandas/main:
mem_x_main

with changes in this PR:
mem_x_pr

On Windows 10 (python 3.8.5):
pandas/main:
mem_win_main

with changes in this PR:
mem_win_fix

I run under valgrind (10 loops), less revealing but there're diff:
pandas/main:
``` text
...
==3116== LEAK SUMMARY:
==3116==    definitely lost: 94,024 bytes in 583 blocks
==3116==    indirectly lost: 7,248,888 bytes in 537 blocks
==3116==      possibly lost: 5,900,491 bytes in 40,529 blocks
==3116==    still reachable: 10,492,550 bytes in 67,907 blocks
==3116==                       of which reachable via heuristic:
==3116==                         stdstring          : 2,484 bytes in 62 blocks
==3116==         suppressed: 0 bytes in 0 blocks
==3116== Reachable blocks (those to which a pointer was found) are not shown.
==3116== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==3116== 
==3116== For lists of detected and suppressed errors, rerun with: -s
==3116== ERROR SUMMARY: 6706 errors from 6703 contexts (suppressed: 4 from 4)
```

with changes in this PR:
``` text
...
==3026== LEAK SUMMARY:
==3026==    definitely lost: 92,296 bytes in 565 blocks
==3026==    indirectly lost: 47,232 bytes in 483 blocks
==3026==      possibly lost: 5,099,615 bytes in 40,511 blocks
==3026==    still reachable: 10,492,630 bytes in 67,908 blocks
==3026==                       of which reachable via heuristic:
==3026==                         stdstring          : 2,484 bytes in 62 blocks
==3026==         suppressed: 0 bytes in 0 blocks
==3026== Reachable blocks (those to which a pointer was found) are not shown.
==3026== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==3026== 
==3026== For lists of detected and suppressed errors, rerun with: -s
==3026== ERROR SUMMARY: 6698 errors from 6695 contexts (suppressed: 8 from 4)
```

### CODE REVIEW

**Comment by WillAyd** [comment]
@vernetya great find! I think this will end up making its way into 1.4.1 whenever the whatsnew for that gets created. Code-wise lgtm @jreback @jbrockmendel @simonjayhawkins

File: doc/source/whatsnew/v1.5.0.rst [hunk 1]
```diff
@@ -227,7 +227,7 @@ MultiIndex
 I/O
 ^^^
 - Bug in :meth:`DataFrame.to_stata` where no error is raised if the :class:`DataFrame` contains ``-np.inf`` (:issue:`45350`)
--
+- Bug in :meth:`to_json` fix memory leak (:issue:`43877`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.5.0.rst:230]: @vernetya can you move to 1.4 (if you can do tonight will merge) otherwise can put in 1.4.1

**Comment by jreback** [comment]
@simonjayhawkins not a blocker for release (we can put in 1.4.1 as well)

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @vernetya

**Comment by jreback** [comment]
@meeseeksdev backport 1.4.x

**Comment by vernetya** [comment]
Thanks for the backport

**Pull request merged**